Automattic/wp-calypso
# Staging Sites: Add production to staging sync

Related to https://github.com/Automattic/dotcom-forge/issues/3691

## Proposed Changes

This pr add a prototype for production -> staging sync

## Testing Instructions


### Test 1: UI
1. Create a WoA site
2. Add a staging site
3. While being on the production site, ensure that you see a "Push to staging" button like below:
![2023-09-21T13:22:49,843711223+03:00](https://github.com/Automattic/wp-calypso/assets/497103/33a9c0fc-516c-46af-8688-9615b27ee8cb)
4. Press the button. Ensure that you see the following modal:
![2023-09-22T16:28:27,563841194+03:00](https://github.com/Automattic/wp-calypso/assets/497103/3e7345bd-e731-49ce-bcd0-eea399f57e33)
5. Navigate to the staging side:
6. Ensure that you see a "Pull from production" button.
![2023-09-21T13:23:25,028884515+03:00](https://github.com/Automattic/wp-calypso/assets/497103/2406fabb-50d8-4919-bc6b-93dbdb77e98a)
7. Press the button. Ensure that you see the following modal:
![2023-09-21T13:23:38,943854376+03:00](https://github.com/Automattic/wp-calypso/assets/497103/6510d099-f583-4f8f-8030-b37a9590fdf0)

### Test 2: Functionality
1. Create a WoA site
2. Add a staging site
3. Apply D122836-code.
4. At your production site, change your theme ( eg: `Awburn` ).
5. While on the production side, press the push button. Ensure that the endpoint called is `/wpcom/v2/sites/<prod_id>/staging-site/push-to-staging/<staging_id>`
6. While on the staging side, press the pull button. Ensure that the endpoint called is `/wpcom/v2/sites/<staging id>/staging-site/pull-from-production/<production_id>`
7. Navigate to rewind debugger. Ensure that you see a new queued job.
8. Wait a bit. For me, it took several minutes (10-15). Ensure that your site has been restored successfully.
9. Navigate to `Appearance -> Themes` at your staging site and ensure that your theme has been changed.

### Test 3: Analytics
1. Create a WoA site
2. Add a staging site
3. Apply D122836-code
4. Change the code to return an error.
5. While on the production side, press the push button. Ensure that you see `calypso_hosting_configuration_staging_site_push_failure` event to being dispatched.
6. While on the staging side, press the pull button.  Ensure that you see `calypso_hosting_configuration_staging_site_pull_failure` being dispatched.

## Pre-merge Checklist
